docs/enhancements/infrastructure-monitoring-service-api/alarms.md

```diff
@@ -0,0 +1,669 @@
+
+
+-- Update the sequence if resolved or alarm_changed_time changed 
+CREATE OR REPLACE FUNCTION set_alarm_sequence_on_update()
+    RETURNS TRIGGER AS $$
+BEGIN
+    IF (NEW.status = 'resolved' AND OLD.status IS DISTINCT FROM 'resolved')
+        OR (NEW.alarm_changed_time IS DISTINCT FROM OLD.alarm_changed_time) THEN
+        NEW.alarm_sequence_number := nextval('alarm_sequence_seq');
+    END IF;
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+
+-- Attach the trigger to alarm_event_record
+CREATE TRIGGER trg_alarm_sequence_update
+    BEFORE UPDATE ON alarm_event_record
+    FOR EACH ROW
+EXECUTE FUNCTION set_alarm_sequence_on_update();
+
+
+-- Table: alarm_event_record_archive, is identical to alarm_event_record. Use to this to eventually store events that are considered historical (i.e status: resolved)
+DROP TABLE IF EXISTS alarm_event_record_archive CASCADE;
+CREATE TABLE alarm_event_record_archive
+(LIKE alarm_event_record INCLUDING ALL);
+```
+
+Please note that script above was used test and need some updates for production e.g use `CREATE TABLE IF NOT EXISTS` and remove `DROP TABLE IF EXISTS`. 
+Apply this during deployment as specified [here](#k8s-resources) via a migration tool as specified [here](#tooling-and-general-dev-guidelines). 
+
+## Init
+Notes: 
+- Interacting with the following: `versions`, `alarm_dictionary` and `alarm_definitions`
+- `alarm_dictionary` has a primary key `resource_type_id`, ensuring that each ResourceType has a unique AlarmDictionary.
+- There's a one-to-many relation between `alarm_dictionary` and `alarm_definitions` via major.minor OCP version. The referential integrity and normalization version is done through `versions` table
+
+```sql
+INSERT INTO versions (version_number)
+VALUES
+    ('4.16'),
+    ('4.17');
+
+-- Most of the data in this table will come from k8s calls for now
+INSERT INTO alarm_dictionary (alarm_dictionary_version, resource_type_id, entity_type, vendor)
+VALUES
+    ('4.16', 'b3e7149e-d471-4d0f-aaa6-d5e9aa9e713a', 'telco-model-OpenShift-4.16.2', 'Red Hat'),
+    ('4.16', '481688c8-2782-4534-a9de-88ca5154411d', 'telco-model-OpenShift-4.16.8', 'Red Hat' ),
+    ('4.17', 'f8b9e100-fd9f-4923-b96f-89418e9c2560', 'telco-model-OpenShift-4.17.2', 'Red Hat');
+
+-- Insert into alarm_definitions
+INSERT INTO alarm_definitions (alarm_name, alarm_last_change, alarm_description, proposed_repair_actions , alarm_additional_fields)
+VALUES
+    ('NodeClockNotSynchronising','4.16', 'Clock not synchronising.','Clock at {{ $labels.instance }} is not synchronising. Ensure NTP is configured on this host.','{"CustomKey": "CustomValue"}'),
+    ('LowMemory','4.16','Low memory.','Low memory, with a longer description to help user fix the issue','{"CustomKey2": "CustomValue"}'),
+    ('NodeClockNotSynchronising','4.17','Clock not synchronising.','Clock at {{ $labels.instance }} is not synchronising. Ensure NTP2 is configured on this host.', '{"CustomKey3": "CustomValue"}');
```

Comment:
Where does this data come from? How will we decide which alarms to define? Where does the proposed_repair_action come from?

Reply:
> Where does this data come from
Each definition is a Rule defined in PrometheusRule CR 

> How will we decide which alarms to define
We are retrieving all the  PrometheusRule CRs from unique major.minor clusters. I mentioned it in the doc but e.g if we are managing 3 clusters 4.16.2, 4.17.2 and 4.16.8, Pick 4.16.8 and 4.17.2 which effectively represents all the rules in 4.16.z and 4.17.z clusters

> Where does the proposed_repair_action come from
I changed the doc a bit more to understand the mapping (##PrometheusRule to AlarmDefinition mapping) but basically this is `rules.annotations.runbook_url` if available

Reply:
Please update the section above to reflect your comments

Reply:
Updated the example with real data